src/tools/idlc/src/gen_c99.c

```diff
@@ -0,0 +1,2063 @@
+      if (rc != DDS_RETCODE_OK) {
+        return rc;
+      }
+    }
+  }
+
+  return DDS_RETCODE_OK;
+}
+
+/* Finding specific parent of a type */
+
+static ddsts_type_t *ddsts_type_get_parent(ddsts_type_t *type, ddsts_flags_t flags)
+{
+  while (type != NULL) {
+    type = type->type.parent;
+    if (type != NULL && DDSTS_IS_TYPE(type, flags)) {
+      return type;
+    }
+  }
+  return NULL;
+}
+
+/* Output function with named string arguments */
+
+static void output(ddsts_ostream_t *ostream, const char *fmt, ...)
+{
+  const char *values[26];
+  for (int i = 0; i < 26; i++) {
+    values[i] = "";
+  }
+
+  /* Parse variable arguments: pairs of a capital letter and a string */
+  va_list args;
+  va_start(args, fmt);
+  for (;;) {
+    char letter = (char)va_arg(args, int);
+    if (letter < 'A' || letter > 'Z') {
+      break;
+    }
+    values[letter - 'A'] = va_arg(args, const char*);
+  }
+  va_end(args);
+
+  const char *s;
+  for (s = fmt; *s != '\0'; s++) {
+    if (*s == '$') {
+      s++;
+      if (*s == '$')
+        ddsts_ostream_put(ostream, *s);
+      else if ('A' <= *s && *s <= 'Z') {
+        const char *t;
+        for (t = values[*s - 'A']; *t != '\0'; t++) {
+          ddsts_ostream_put(ostream, *t);
+        }
+      }
+    }
+    else {
+      ddsts_ostream_put(ostream, *s);
+    }
+  }
+}
+
+
+/* File name functions */
+
+static dds_retcode_t output_file_name(const char *file_name, const char *ext, const char **ref_result)
+{
+  *ref_result = NULL;
+  size_t file_name_len = strlen(file_name);
+  if (file_name_len < 4 || strcmp(file_name + file_name_len - 4, ".idl") != 0) {
+    DDS_ERROR("Error: File name '%s' does not have '.idl' extension", file_name);
+    return DDS_RETCODE_ERROR;
+  }
+  size_t result_len = file_name_len - 2 + strlen(ext);
+  char *result = (char*)ddsrt_malloc(sizeof(char)*(result_len));
+  if (result == NULL) {
+    return DDS_RETCODE_OUT_OF_RESOURCES;
+  }
+  ddsrt_strlcpy(result, file_name, file_name_len - 2);
+  ddsrt_strlcat(result, ext, result_len);
+  *ref_result = result;
+  return DDS_RETCODE_OK;
+}
+
+static dds_retcode_t uppercase_file_name(const char *file_name, const char **ref_result)
+{
+  *ref_result = NULL;
+  size_t file_name_len = strlen(file_name);
+  if (file_name_len < 4 || strcmp(file_name + file_name_len - 4, ".idl") != 0) {
+    DDS_ERROR("Error: File name '%s' does not have '.idl' extension", file_name);
+    return DDS_RETCODE_ERROR;
+  }
+  char *result = (char*)ddsrt_malloc(sizeof(char)*(file_name_len - 3));
+  if (result == NULL) {
+    return DDS_RETCODE_OUT_OF_RESOURCES;
+  }
+  size_t i;
+  for (i = 0; i < file_name_len - 4; i++)
+    result[i] = (char)toupper(file_name[i]);
+  result[i] = '\0';
+  *ref_result = result;
+  return DDS_RETCODE_OK;
+}
+
+/* Generate copyright headere */
+
+static void write_copyright_header(ddsts_ostream_t *ostream, const char *target_file_name, const char *source_file_name)
```

Comment:
I propose to rename this function as it doesn't actually write any copyright information and it might give people a scare. That does remind me that we may have to include something that says code generated by the compiler is licensed under X or something? @eboasson, any thoughts?

Reply:
My personal view is that the output should not contain any copyright claim or license, it is only a mechanically transformed version of the input with a pretty straightforward and mostly standardized mapping. Moreover, we cannot know what the input copyright/license is, and therefore a copyright/license statement in the output would be problematic at best.

Reply:
I change the name into write_file_header. The function used to print some copyright message, but it already did not do that anymore.